src/vehicle.cpp

```diff
@@ -2595,8 +2621,10 @@ int vehicle::total_mass()
           continue;
         }
         m += item::find_type( part_info(i).item )->weight;
-        for( auto &j : get_items(i) ) {
-            m += j.type->weight;
+        if( part_flag( i, "CARGO" ) ) {
+            for( auto &j : get_items(i) ) {
+                m += j.type->weight;
+            }
```

Comment:
If you keep the original loop, but use `j.weight()`, the calculation should be fine.
`item::weight` returns the correct values for item counted by charges (which would apply to the stored liquid).

The current usage of just the item types weight is incorrect anyway. All items with charges will weight the same!